EKF/vel_pos_fusion.cpp

```diff
@@ -45,276 +45,174 @@
-	float gate_size[6] = {}; // innovation consistency check gate sizes for [VN,VE,VD,PN,PE,PD] observations
-	float Kfusion[24] = {}; // Kalman gain vector for any single observation - sequential fusion is used
-	float innovation[6]; // local copy of innovations for  [VN,VE,VD,PN,PE,PD]
-	memcpy(innovation, _vel_pos_innov, sizeof(_vel_pos_innov));
-
-	// calculate innovations, innovations gate sizes and observation variances
-	if (_fuse_hor_vel || _fuse_hor_vel_aux) {
-		// enable fusion for NE velocity axes
-		fuse_map[0] = fuse_map[1] = true;
-
-		// handle special case where we are getting velocity observations from an auxiliary source
-		if (!_fuse_hor_vel) {
-			innovation[0] = _aux_vel_innov[0];
-			innovation[1] = _aux_vel_innov[1];
-		}
-
-		// Set observation noise variance and innovation consistency check gate size for the NE position observations
-		R[0] = _velObsVarNED(0);
-		R[1] = _velObsVarNED(1);
-		gate_size[1] = gate_size[0] = _hvelInnovGate;
 
+bool Ekf::fuseHorizontalVelocity(const Vector3f &innov, const Vector2f &innov_gate,
+				 const Vector3f &obs_var, Vector3f &innov_var, Vector2f &test_ratio)
+{
+	innov_var(0) = P[4][4] + obs_var(0);
+	innov_var(1) = P[5][5] + obs_var(1);
+	test_ratio(0) = fmaxf( sq(innov(0)) / (sq(innov_gate(0)) * innov_var(0)),
+			       sq(innov(1)) / (sq(innov_gate(0)) * innov_var(1)));
+
+	bool innov_check_pass = (test_ratio(0) <= 1.0f);
+	if (innov_check_pass)
+	{
+		_time_last_hor_vel_fuse = _time_last_imu;
+		_innov_check_fail_status.flags.reject_hor_vel = false;
+
+		fuseVelPosHeight(innov(0),innov_var(0),0);
+		fuseVelPosHeight(innov(1),innov_var(1),1);
+
+		return true;
+	}else{
+		_innov_check_fail_status.flags.reject_hor_vel = true;
+		return false;
 	}
+}
 
-	if (_fuse_vert_vel) {
-		fuse_map[2] = true;
-		// observation variance - use receiver reported accuracy with parameter setting the minimum value
-		R[2] = _velObsVarNED(2);
-		// use scaled horizontal speed accuracy assuming typical ratio of VDOP/HDOP
-		R[2] = 1.5f * fmaxf(R[2], _gps_sample_delayed.sacc);
-		R[2] = R[2] * R[2];
-		// innovation gate size
-		gate_size[2] = _vvelInnovGate;
-	}
+bool Ekf::fuseVerticalVelocity(const Vector3f &innov, const Vector2f &innov_gate,
+				 const Vector3f &obs_var, Vector3f &innov_var, Vector2f &test_ratio)
+{
+	innov_var(2) = P[6][6] + obs_var(2);
+	test_ratio(1) = sq(innov(2)) / (sq(innov_gate(1)) * innov_var(2));
 
-	if (_fuse_pos) {
-		// enable fusion for the NE position axes
-		fuse_map[3] = fuse_map[4] = true;
+	bool innov_check_pass = (test_ratio(1) <= 1.0f);
+	if (innov_check_pass) {
+		_time_last_ver_vel_fuse = _time_last_imu;
+		_innov_check_fail_status.flags.reject_ver_vel = false;
 
-		// Set observation noise variance and innovation consistency check gate size for the NE position observations
-		R[4] = R[3] = sq(_posObsNoiseNE);
-		gate_size[4] = gate_size[3] = _posInnovGateNE;
+		fuseVelPosHeight(innov(2),innov_var(2),2);
 
+		return true;
+	}else{
```

Comment:
```suggestion
	} else {
```

Reply:
That is only the tip of the iceberg. At some point we really need a formatter.